compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -638,15 +640,31 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
       }
     }
 
+    // Fix subtype checking for child instantiation,
+    // such that `Foo(Test.this.foo) <:< Foo(Foo.this)`
```

Comment:
I could not correlate the claim `Foo(Test.this.foo) <:< Foo(Foo.this)` in i3939.scala with the code. In i3938, foo is not a module, so how does the `removeThisType` affect it?

Also, note that if we do have a 

     object Foo

then we have already `Foo.type <:< Foo.this` and `Foo.this <:< Foo.type` by the rules for `ThisType` in TypeComparer's firstTry and secondTry methods. So I am sure why the code would be necessesary.

Reply:
Thanks for the clarification. The correlation is that in the `else` branch, it throws away of the `ThisType`. For the `module` case, I'll try if I can throw it away.

Reply:
I just checked, the patch for `module` cannot be removed, as the subtyping for `Foo.type <:< Foo.this` only works if it's static owner:

```Scala
          case tp1: NamedType if cls2.is(Module) && cls2.eq(tp1.widen.typeSymbol) =>
            cls2.isStaticOwner ||
            isSubType(tp1.prefix, cls2.owner.thisType) ||
            secondTry(tp1, tp2)
```

The test case `tests/patmat/3455.scala` shows that we need the patch:

```Scala
trait AxisCompanion {
   sealed trait Format
   object Format {
      case object Decimal extends Format
      case object Integer extends Format
   }
}
object Axis extends AxisCompanion
class Axis {
   import Axis._
   def test( f: Format ) = f match {
      case Format.Integer => "Int"
   }
}
```